DDFileLogger queues and refactoring

hhanesand opened the pull request:
- Use already provided queue to manage access to variables in file logger. The queue is created in the abstract superclass as a serial queue, which fits our needs.
- Prefix internal methods meant to be called on internal queue and add asserts to avoid further threading issues.
- Refactor out some if statements determining which log files to archive.
- Add DISPATCH_VNODE_REVOKE to vnode observation flags.
- Remove usage of `fileExistsAtPath:`, instead just attempt the operation. withIntermediateDirectories: YES doesn't return error if directory exists.

hhanesand commented on Classes/DDFileLogger.h:
```diff
@@ -302,9 +302,7 @@ extern unsigned long long const kDDDefaultLogFilesDiskQuota;
-@interface DDFileLogger : DDAbstractLogger <DDLogger> {
-	DDLogFileInfo *_currentLogFileInfo;
-}
+@interface DDFileLogger : DDAbstractLogger <DDLogger>
```
This absolutely has to be protected by the logger queue, and hence we can't expose it here in the header. It's been public since 2016, so this is a breaking change.

hhanesand commented on Classes/DDFileLogger.m:
```diff
@@ -263,15 +263,13 @@ - (NSString *)logsDirectory {
+                                                              attributes:nil
+                                                                   error:&err];
+    if (success == NO) {
+        NSLogError(@"DDFileLogManagerDefault: Error creating logsDirectory: %@", err);
```
For this, see the comment above `-[NSFileManager fileExistsAtPath]`

hhanesand replied:
when withIntermediateDirectories:YES is specified, the FileManager does not return any errors if the directory already exists.

Doc for return value on that method :
> true if the directory was created, true if createIntermediates is set and the directory already exists, or false if an error occurred.

https://developer.apple.com/documentation/foundation/filemanager/1407884-createdirectory

ffried replied:
Ah, you're right. Thanks for the heads up... 🙂

OSSPolice commented:
<!--
  0 Errors
  2 Warnings: Any changes to library code sh...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any changes to library code should be reflected in the CHANGELOG. Please consider adding a note there about your change.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">The library files were changed, but the tests remained unmodified. Consider updating or adding to the tests to match the library changes.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hhanesand commented on Classes/DDFileLogger.m:
```diff
@@ -827,23 +812,21 @@ - (void)rollLogFileNow {
+        if ([self->_logFileManager respondsToSelector:@selector(didRollAndArchiveLogFile:)]) {
+            [self->_logFileManager didRollAndArchiveLogFile:archivedFilePath];
+        }
+    });
```
adding a dispatch_async here prevents deadlock if _logFileManager logs an error

hhanesand commented on Classes/DDFileLogger.m:
```diff
@@ -884,128 +870,153 @@ - (void)maybeRollLogFileDueToSize {
-
-                self->_currentLogFileInfo = [[DDLogFileInfo alloc] initWithFilePath:currentLogFilePath];
-            }
+            });
```
adding a dispatch_async here prevents deadlock if _logFileManager logs an error

hhanesand commented on Classes/DDFileLogger.h:
```diff
@@ -302,9 +306,7 @@ extern unsigned long long const kDDDefaultLogFilesDiskQuota;
  */
-@interface DDFileLogger : DDAbstractLogger <DDLogger> {
-	DDLogFileInfo *_currentLogFileInfo;
-}
```
This absolutely has to be protected by the logger queue, and hence we can't expose it here in the header. It's been public since 2016, so this is a breaking change.

ffried commented on Classes/DDFileLogger.m:
```diff
@@ -905,128 +933,152 @@ - (void)maybeRollLogFileDueToSize {
-                
+- (void)lt_monitorCurrentLogFileForExternalChanges {
+    NSAssert(self->_currentLogFileHandle, @"Can not monitor without handle.");
+    NSAssert([self isOnInternalLoggerQueue], @"lt_ methods should be on logger queue.");
```
Nit-Pick: Logically we should first assert being on the internal queue and then access the currentLogFileHandle.

hhanesand replied:
good catch

ffried commented on Classes/DDFileLogger.m:
```diff
@@ -1044,18 +1096,24 @@ - (void)logMessage:(DDLogMessage *)logMessage {
 
     if (message) {
-        if ((!isFormatted || _automaticallyAppendNewlineForCustomFormatters) &&
-            (![message hasSuffix:@"\n"])) {
```
Like above, I think we're fine with expecting `\n` to be the newline character here.

ffried commented on Classes/DDFileLogger.m:
```diff
@@ -1077,27 +1135,26 @@ - (void)willLogMessage {
 
-    [self rollLogFileNow];
+- (void)flush {
+    NSAssert([self isOnInternalLoggerQueue], @"lt_ methods should be on logger queue.");
```
Assertion method is incorrect, it's not an `lt_` method...

hhanesand replied:
This method is currently only called from inside `DDLog lt_flush`, which dispatches onto the logger's queue before calling the method. I was considering renaming it, but I believe that would be another breaking change since the method is defined in the public `DDLogger` protocol. I therefore opted to insert the check and not rename the method. I'll insert a comment since this is rather non-intuitive.

ffried replied:
A comment would most certainly help, yea.
But why deprecate `flush` in your alternative? Why not just add `lt_flush` and forward `flush` -> `lt_flush`? People calling `flush` from outside the logger probably *want* it to be safe (as in dispatched to the internal queue), right?

hhanesand replied:
Yea, let's make flush save for outside use 👍

ffried commented on Classes/DDFileLogger.m:
```diff
@@ -391,7 +402,8 @@ - (NSArray *)sortedLogFileNames {
 
 - (NSArray *)sortedLogFileInfos {
-    return  [[self unsortedLogFileInfos] sortedArrayUsingComparator:^NSComparisonResult(DDLogFileInfo   * _Nonnull obj1, DDLogFileInfo   * _Nonnull obj2) {
+    return  [[self unsortedLogFileInfos] sortedArrayUsingComparator:^NSComparisonResult(DDLogFileInfo *obj1,
```
Nit-Pick: Two spaces after `return`.

hhanesand commented on Classes/DDFileLogger.h:
```diff
@@ -310,19 +316,23 @@ extern unsigned long long const kDDDefaultLogFilesDiskQuota;
 - (instancetype)init;
 
+- (instancetype)initWithLogFileManager:(id <DDLogFileManager> __nullable)logFileManager
+        __attribute__((deprecated("Use -initWithLogFileManager:completionQueue:")));
```
I was worried about visibility here. The warning would point out the change and introduce the new initializer. What do you think? I'm open to both.

ffried replied:
I know many developers (me included) who work with the "Treat warnings as errors" compiler setting enabled. Those will hate us for this. 😉 
Also, those that really need this new initializer will find it. Those who don't, anyways don't care and are happy if there's neither a warning nor an error.
But what you can do, to improve visibility is to add a header doc to this initializer. People working with Xcode's quick help will see it, as well as those looking at the headers.

hhanesand replied:
Went ahead and added the header docs.

hhanesand commented on Classes/DDFileLogger.m:
```diff
@@ -429,32 +442,46 @@ - (NSString *)newLogFileName {
-    
-    NSData *fileHeaderData = [fileHeaderStr dataUsingEncoding:NSUTF8StringEncoding];
-    return fileHeaderData;
+    unichar lastChar = [fileHeaderStr characterAtIndex:fileHeaderStr.length - 1];
```
yeah, you're right. The logFileHeader shouldn't contain any other newlines

hhanesand commented on Classes/DDFileLogger+Internal.h:
```diff
@@ -18,8 +18,9 @@
+
+- (void)lt_logData:(NSData *)data;
+
 @end
```
Saw no reason to have logData calling writeToFile, when writeToFile is not used anywhere else in the project.

hhanesand commented on Classes/Extensions/DDFileLogger+Buffering.m:
```diff
@@ -64,15 +70,14 @@ - (void)appendToBuffer:(NSData *)data {
-        if (appendedData != NULL) {
+            _bufferSize += length;
+
             free((void *)appendedData);
```
Made a slight semantic change here.

hhanesand commented:
@ffried I've addressed comments from your review and merged master. Looked over the buffering code and it should play fine with the queue changes. I'll test this branch in one of my projects today.

ffried commented:
@hhanesand thanks! I’ll take another look at the code tomorrow. Once you give the green light after your tests, we can merge this. 👍
